### PR TITLE
[II] Serve Kimi-K3 QSRT on TP16 with DCP

### PR DESCRIPTION
**Pull request by voipmonitor**
## Behavior

- Serves the official `moonshotai/Kimi-K3` MXFP4 checkpoint and the tensor-parallel-independent `lukealonso/Kimi-K3-QSRT-K2` checkpoint on 16 GPUs through B12X dense MLA, routed MoE, linear kernels, and PCIe collectives.
- Supports DCP4, DCP8, and DCP16 cache layouts without padding Kimi-K3's 896 routed experts.
- Supports target-only decode, Inferact DSpark, and DFlash with graph-owned B12X projection pools and caller-owned dense-MLA scratch.
- Preserves causal and non-causal speculative metadata, DCP-local cache planning, InstantTensor borrowed-buffer ownership, exact routed-expert selection, and bounded adaptive DSpark depth.
- Adds `VLLM_MLA_CHUNKED_PREFILL_WORKSPACE_SIZE` so transient dense MLA K/V projections can be bounded independently of maximum context length. Zero retains the 64K-token automatic bound; negative values fail configuration.

## Technical reason

Kimi-K3's routed-expert count and MLA projection shapes do not divide uniformly across TP16. The implementation shards compatible projections, gathers only required projection rows, preserves canonical expert identities, and uses DCP-aware attention metadata. A separately bounded prefill workspace prevents a 1,048,576-token model limit from reserving multi-gigabyte transient dense K/V projections on every GPU.

## Compatibility

Kimi-K3 behavior is selected by model type and explicit environment variables. TP12/DCP1 QSRT serving remains supported. Process groups without B12X projection pools retain no-op graph-capture behavior. The MLA workspace retains automatic sizing unless `VLLM_MLA_CHUNKED_PREFILL_WORKSPACE_SIZE` is nonzero.

The B12X narrow-output, paired top-k, and vocabulary-argmax dependency is [local-inference-lab/b12x#198](https://github.com/local-inference-lab/b12x/pull/198). Merge B12X #198 before using the optimized vLLM paths.

The official-MXFP4 integration proposed by #284 is included in this pull request against `dev/infernal-invocation`; #284 is superseded. PRs #243, #269, and #310 target different checkpoint families, model branches, or draft-model semantics and are not duplicates.

## Validation

Validation used 16 NVIDIA RTX PRO 6000 Blackwell Workstation Edition GPUs, CUDA 13.3, PyTorch 2.13.0, FP8 target KV cache, and InstantTensor loading.

- Ruff and Python compilation pass for every changed Python file.
- Focused CPU and GPU-visible suites pass 54 tests with 11 hardware-inapplicable skips. Their GPU coverage includes 12 MLA context tests plus DCP chunking, B12X graph-pool ownership and rollback, InstantTensor lifecycle, reload behavior, speculative attention metadata, sharded Markov execution, fused top-k, and Kimi-K3 kernel warmup.
- An additional integration-invariant suite passes 12 CPU tests with 9 CUDA-only skips; it covers independent draft quantization, complete K-step sharded Markov all-reduce sizing, hierarchical B12X runtime discovery, and fused-router warmup without a KDA layer.
- The FA4 backport applies idempotently across repeated CMake patch steps at `vllm-project/flash-attention@f3e1a4f74c99145c0717709860bf765de1703779` and produces wrapper SHA-256 `80058d9ea24fb51eaf1edecf4413d9c2008b0a5538127263a2e2500e8f838479`.
- Official MXFP4 TP16/DCP16 target-only, DSpark K7, and DFlash K7 expose at least 1,048,576 physical target-cache tokens. Their one-request median decode rates are 56.05, 118.92, and 89.35 tok/s, respectively.
- QSRT K2 TP16/DCP4 and TP16/DCP8 pass a 16-context finite-logit suite and coherent coding generation.
- QSRT K2 TP16/DCP8 target-only decode reaches a 47.27 tok/s median with a 256-token prompt and 384 generated tokens.
- QSRT K2 TP16/DCP8 with static seven-token DSpark exposes at least 1,048,576 physical target-cache tokens and reaches a 91.23 tok/s median across five identical 1,024-token outputs.
- A TP16/DCP8 QSRT K2 server with a 4,096-token scheduler budget exposes 1,102,812 physical target-cache tokens while serving 71,136-114,776-token AA-LCR prompts with eight active sequences.

## Source identity

- Pull-request head: `3a50cc050cf70ad3be0f862d361e5fc0a3ec730a`
- Resulting Git tree: `735952bd5d9703b7fe966fe94f9e570f13169e92`
- Base: `dev/infernal-invocation@ad848fc4141f201489db18d5453c50b312245a0a`

## AI assistance

Codex composed the implementation, generated tests and validation commands, and inspected the resulting diff and runtime evidence. Human review remains required before merge.

### CODE REVIEW

File: serve-kimi-k3-full-mxfp4-dcp16-1m-dflash.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+export SERVED_MODEL_NAME="${SERVED_MODEL_NAME:-Kimi-K3-MXFP4-DFlash-DCP16-1M}"
+export MAX_MODEL_LEN="${MAX_MODEL_LEN:-1048576}"
+export KV_CACHE_MEMORY_BYTES="${KV_CACHE_MEMORY_BYTES:-1200000000}"
+export VLLM_USE_B12X_FP8_GEMM="${VLLM_USE_B12X_FP8_GEMM:-1}"
+export VLLM_KIMI_FUSED_TOPK16="${VLLM_KIMI_FUSED_TOPK16:-1}"
+export VLLM_MEMORY_PROFILER_ESTIMATE_CUDAGRAPHS="${VLLM_MEMORY_PROFILER_ESTIMATE_CUDAGRAPHS:-1}"
+# The target and DFlash capture sequences require distinct MoE output storage.
+export VLLM_DISABLE_FUSED_MOE_OUTPUT_ALIAS="${VLLM_DISABLE_FUSED_MOE_OUTPUT_ALIAS:-1}"
+export GLOO_SOCKET_IFNAME="${GLOO_SOCKET_IFNAME:-lo}"
+export TP_SOCKET_IFNAME="${TP_SOCKET_IFNAME:-lo}"
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply serve-kimi-k3-full-mxfp4-dcp16-1m-dflash.sh:40]: Fixed in 1dd3cab17e. The DFlash launcher now derives VLLM_DCP_A2A_MAX_TOKENS from DFLASH_TOKENS + 1 while preserving an explicit user override.

File: tests/model_executor/model_loader/instanttensor_loader/test_weight_utils.py [hunk 2]
```diff
@@ -160,6 +164,75 @@ def test_instanttensor_restricts_io_to_indexed_shards(tmp_path):
...
+    save_file(
+        {
+            "model.large.weight": torch.arange(8, dtype=torch.float32),
+            "model.small.weight": torch.arange(2, dtype=torch.float32),
+        },
+        shard,
+    )
+    physical_names = []
+    with weight_utils.safe_open(shard, framework="pt") as physical_file:
+        physical_names = list(physical_file.offset_keys())
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tests/model_executor/model_loader/instanttensor_loader/test_weight_utils.py:182]: No code change is required for this finding. The pinned runtime uses safetensors 0.8.0, whose safetensors.safe_open handle provides offset_keys; the API is also present in safetensors 0.7.0. This helper intentionally reads physical header order from safetensors before rewriting the unopened InstantTensor metadata. Opening the shard through InstantTensor here would duplicate the GPU loader and would not provide the canonical physical-header ordering contract. The InstantTensor unit suite passes: 5 passed, 2 CUDA-only tests skipped.

File: tools/kimi_k3/benchmark_nospec_decode.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import statistics
+import time
+import urllib.request
+from pathlib import Path
+from typing import Any
+
+
+def _get(url: str) -> bytes:
+    with urllib.request.urlopen(url, timeout=30) as response:
+        return response.read()
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tools/kimi_k3/benchmark_nospec_decode.py:38]: Fixed in 1dd3cab17e. The benchmark validates an HTTP or HTTPS scheme and a nonempty authority before either discovery or completion requests; the narrow S310 suppressions are adjacent to those validated calls.

File: tools/kimi_k3/benchmark_nospec_decode.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    decode_seconds = token_times[-1] - token_times[0]
+    return {
+        "formula": "(completion_tokens - 1) / (last_token_time - first_token_time)",
+        "prompt_tokens": int(usage.get("prompt_tokens", len(prompt))),
+        "completion_tokens": completion_tokens,
+        "allowed_token_id": allowed_token_id,
+        "timed_events": len(token_times),
+        "ttft_seconds": token_times[0] - started,
+        "decode_seconds": decode_seconds,
+        "decode_tokens_per_second": (completion_tokens - 1) / decode_seconds,
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tools/kimi_k3/benchmark_nospec_decode.py:122]: Fixed in 1dd3cab17e. Throughput is calculated only after the streamed choice-event count equals usage.completion_tokens, so the numerator and timestamps describe the same token sequence.

File: vllm/envs.py [hunk 5]
```diff
@@ -2343,6 +2392,30 @@ def _resolve_rust_cli_path() -> str | None:
     "VLLM_MEMORY_PROFILE_INCLUDE_ATTN": lambda: bool(
         int(os.getenv("VLLM_MEMORY_PROFILE_INCLUDE_ATTN", "0"))
     ),
+    # TP-shard Kimi-K3's merged MLA q_a/kv_a projection. The rank-major
+    # result is reconstructed losslessly before the two logical outputs are
+    # consumed.
+    "VLLM_KIMI_SHARD_QKV_A": lambda: bool(int(os.getenv("VLLM_KIMI_SHARD_QKV_A", "0"))),
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/envs.py:2398]: Fixed in 1dd3cab17e. The Kimi environment loaders are wrapped to the repository line-length convention without changing their values.

File: vllm/model_executor/models/qwen3_dspark.py [hunk 6]
```diff
@@ -57,6 +59,30 @@ def __init__(
...
+                    prefix=maybe_prefix(prefix, "markov_w1"),
+                )
+            )
+            self.markov_w2 = ParallelLMHead(
+                draft_vocab_size,
+                markov_rank,
+                bias=False,
+                prefix=maybe_prefix(prefix, "markov_w2"),
+            )
+            return
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/model_executor/models/qwen3_dspark.py:86]: Fixed in 1dd3cab17e. The TP-sharded Markov W2 constructor now receives quant_config, matching the replicated branch.

File: vllm/models/kimi_k3/nvidia/model.py [hunk 7]
```diff
@@ -602,6 +627,75 @@ def make_kimi_k3_mega_moe_expert_params_mapping(
...
+        if (
+            self.top_k == 16
+            and self.global_num_experts == 896
+            and router_logits.ndim == 2
+            and router_logits.shape == (num_tokens, 32)
+            and router_logits.dtype == torch.float32
+        ):
+            topk_weights = router_logits[:, :16]
+            topk_ids = router_logits[:, 16:].view(torch.int32)
+            return topk_weights, topk_ids
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/models/kimi_k3/nvidia/model.py:698]: Fixed in 1dd3cab17e. Both compact routing-payload decoders now require the requested index type to be None or torch.int32 before reinterpreting the FP32 ID plane.

File: vllm/models/kimi_k3/nvidia/model.py [hunk 8]
```diff
@@ -875,11 +993,48 @@ def _router(
...
+            )
+            if fused_pair_topk is not None:
+                routed_hidden_states, routing_payload = fused_pair_topk
+                return routed_hidden_states, routing_payload, None
+            routed_hidden_states, router_logits = gather_kimi_sharded_projection_pair(
+                down_local,
+                router_local,
+            )
+            router_output, topk_ids = _finish_router(router_logits)
+            return routed_hidden_states, router_output, topk_ids
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/models/kimi_k3/nvidia/model.py:1047]: Fixed in 1dd3cab17e. MegaMoE is excluded from the paired compact-payload path and retains its grouped-routing implementation.

File: vllm/models/kimi_k3/nvidia/tp_projection.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    dcp_group = get_dcp_group()
+    if dcp_group.world_size == tp_size:
+        return dcp_group
+    tp_group = get_tp_group()
+    if tp_group.world_size != tp_size:
+        raise RuntimeError(
+            "Kimi projection group does not span tensor-parallel ranks: "
+            f"group={tp_group.world_size}, TP={tp_size}"
+        )
+    return tp_group
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/models/kimi_k3/nvidia/tp_projection.py:44]: Fixed in 1dd3cab17e. A DCP coordinator is selected for projection gathers only when its ordered rank list and world size equal the tensor-parallel coordinator; otherwise the validated TP coordinator is used.

File: vllm/v1/attention/ops/dcp_alltoall.py [hunk 10]
```diff
@@ -198,15 +233,27 @@ def capture_b12x_dcp_a2a(
...
-        for _, pool in matching_pools:
-            stack.enter_context(pool.capture(stream=stream, channel_id=channel_id))
+    if channel_id is None:
+        if matching_pools:
+            raise RuntimeError(
+                "distributed PCIe DCP graph capture requires an explicit "
+                "semantic channel_id"
+            )
         yield
+        return
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/attention/ops/dcp_alltoall.py:250]: Fixed in 1dd3cab17e. A B12X-enabled capture now rejects a missing semantic channel before any pool is created, while generic captures with B12X disabled and no matching pools retain their no-op compatibility path.

File: vllm/v1/attention/ops/dcp_alltoall.py [hunk 11]
```diff
@@ -838,11 +1141,10 @@ def _dcp_a2a_unpack_combine_kernel(
         else:
             weight = tl.exp2(lse_val - global_lse)
         weight = tl.where(weight != weight, 0.0, weight)
-        partial = tl.load(recv_ptr + recv_base + d_offsets * recv_stride_D).to(
-            tl.float32
+        acc += (
+            tl.load(recv_ptr + recv_base + d_offsets * recv_stride_D).to(tl.float32)
+            * weight
         )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/attention/ops/dcp_alltoall.py]: Fixed in 1dd3cab17e. The MLA caller converts locally empty rows to the neutral output/LSE pair (0, -inf) from valid_counts before packed A2A. The PyTorch reference and Triton unpacker also mask zero-weight nonfinite partials before multiplication. This uses metadata already available at the attention call site and does not add unsupported arguments to the public reduction API.

File: vllm/v1/core/kv_cache_utils.py [hunk 12]
```diff
@@ -1181,6 +1181,7 @@ def is_kv_cache_type_attention_free(kv_cache_spec: dict[str, KVCacheSpec]) -> bo
 
 def _get_kv_cache_groups_uniform_page_size(
     kv_cache_spec: dict[str, KVCacheSpec],
+    group_size_override: int | None = None,
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/core/kv_cache_utils.py:1184]: Fixed in 1dd3cab17e. The KV grouping, deferred-loader ownership, online element-count, and InstantTensor range-selection helpers now document their arguments, return values, and applicable exceptions with Google-style sections.

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 13]
```diff
@@ -169,14 +196,24 @@ def _speculative_steps_for_query_len(self, query_len: int) -> int:
...
+            assert self.draft_kv_window_block_size is not None
+            draft_model_config = copy.copy(draft_vllm_config.model_config)
+            draft_model_config.max_model_len = (
+                self.draft_kv_window + self.draft_kv_window_block_size - 1
+            )
+            draft_vllm_config.model_config = draft_model_config
+        draft_attention_config = copy.copy(draft_vllm_config.attention_config)
+        draft_attention_config.use_non_causal = self.requires_non_causal
+        draft_vllm_config.attention_config = draft_attention_config
+        return draft_vllm_config
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:200]: No additional code change is required for this finding. B12xMLAMetadataBuilder.build already treats a source block table wider than plan.caps.max_page_table_width as a flattening condition, copies only the planned prefix into capture-stable storage, and forward_mqa binds that constrained table instead of the target-sized source. test_b12x_mla_builder_bounds_single_token_draft_table covers a width-6 source against a width-4 plan and now states the sequence-length/page-size invariant explicitly. The focused metadata and speculative-decoding suite passes: 24 passed, 9 platform skips.

**Comment by voipmonitor** [comment]
This aggregate change is superseded by the independently reviewable vLLM pull requests #382 through #391.

Those pull requests preserve the required Kimi-K3 behavior as separate units for bounded InstantTensor loading, hybrid KV-cache geometry, DCP collectives, projection sharding, routed-MoE transport, dense MLA, external speculative runtimes, bounded DSpark state, vocabulary-sharded DSpark sampling, and tensor-parallel-specific B12X graph allocation.

The composed vLLM source tree is `f21a391de0a1b127c93ac718fd7d1818f025317b`. It passed 531 tests with 2 skips and 20 explicit deselections. Runtime qualification, merge order, source locks, and the published Docker image are recorded in local-inference-lab/rtx6kpro#66.

Closing #317 prevents the aggregate patch from competing with the reviewable source units.
